testing/trino-product-tests/src/main/java/io/trino/tests/hive/TestHiveTransactionalTable.java

```diff
@@ -1289,6 +1289,410 @@ public void testDeleteWholePartition()
+                        "    WHEN MATCHED AND s.address = 'Centreville' THEN DELETE" +
+                        "    WHEN MATCHED THEN UPDATE SET purchases = s.purchases + t.purchases, address = s.address" +
+                        "    WHEN NOT MATCHED THEN INSERT (customer, purchases, address) VALUES(s.customer, s.purchases, s.address)";
+
+                log.info("About to merge, target table %s, source table %s", targetTable, sourceTable);
+                query(sql);
+
+                log.info("Verifying MERGE");
+                verifySelectForPrestoAndHive("SELECT * FROM " + targetTable, "TRUE", row("Aaron", 11, "Arches"), row("Ed", 7, "Etherville"), row("Bill", 7, "Buena"), row("Dave", 22, "Darbyshire"));
+            });
+        });
+    }
+
+    @Test(groups = HIVE_TRANSACTIONAL, timeOut = TEST_TIMEOUT, dataProvider = "partitionedAndBucketedProvider")
+    public void testMergeUpdateWithVariousLayouts(boolean partitioned, String bucketing)
+    {
+        log.info("In testMergeUpdateWithVariousLayouts, partitioned %s, bucketing %s", partitioned, bucketing);
+        BucketingType bucketingType = bucketing.isEmpty() ? NONE : BUCKETED_V2;
+        withTemporaryTable("merge_with_various_formats", true, true, bucketingType, targetTable -> {
+            StringBuilder builder = new StringBuilder();
+            builder.append("CREATE TABLE ")
+                    .append(targetTable)
+                    .append("(customer STRING");
+            builder.append(partitioned ? ") PARTITIONED BY (" : ", ")
+                    .append("purchase STRING) ");
+            if (!bucketing.isEmpty()) {
+                builder.append(bucketing);
+            }
+            builder.append(" STORED AS ORC TBLPROPERTIES ('transactional' = 'true')");
+            onHive().executeQuery(builder.toString());
+
+            log.info("About to insert");
+            query(format("INSERT INTO %s (customer, purchase) VALUES ('Dave', 'dates'), ('Lou', 'limes'), ('Carol', 'candles')", targetTable));
+            verifySelectForPrestoAndHive("SELECT * FROM " + targetTable, "TRUE", row("Dave", "dates"), row("Lou", "limes"), row("Carol", "candles"));
+
+            withTemporaryTable("merge_simple_source", true, true, NONE, sourceTable -> {
+                query(format("CREATE TABLE %s (customer VARCHAR, purchase VARCHAR) WITH (transactional = true)", sourceTable));
+
+                log.info("Inserting into source");
+                query(format("INSERT INTO %s (customer, purchase) VALUES ('Craig', 'candles'), ('Len', 'limes'), ('Joe', 'jellybeans')", sourceTable));
+
+                String sql = format("MERGE INTO %s t USING %s s ON (t.purchase = s.purchase)", targetTable, sourceTable) +
+                        "    WHEN MATCHED AND s.purchase = 'limes' THEN DELETE" +
+                        "    WHEN MATCHED THEN UPDATE SET customer = CONCAT(t.customer, '_', s.customer)" +
```

Comment:
Do you have a test that would update the partition key (purchase here)?

Reply:
Several of the tests update partition and bucket columns.  For example, `testMergeSimpleSelectPartitioned` updates the `address` partition column at: https://github.com/trinodb/trino/pull/7386/commits/ccc584f6e62ce3f34b1bd3c1c20130486533c8cd#diff-ee4c5c10f12500119c780c8c1ba6b88df410f13b0282bd69e315fe32c5938ffaR1336

However, I love to have suggestions for more tests to write.  I'm always happy to write more tests, but lack imagination as to what would be good ones 😞